File: python/leetcode_grind/day_578.py
# https://leetcode.com/problems/ipo/description/
class Solution1:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        n = len(profits)
        projects = list(zip(capital, profits))
        projects.sort()
        q = []
        ptr = 0
        for i in range(k):
            while ptr < n and projects[ptr][0] <= w:
                heappush(q, -projects[ptr][1])
                ptr += 1
            if not q:
                break
            w += -heappop(q)
        return w

# https://leetcode.com/problems/put-boxes-into-the-warehouse-i/description/
class Solution2:
    def maxBoxesInWarehouse(self, boxes: List[int], warehouse: List[int]) -> int:
        i = 0
        count = 0
        boxes.sort(reverse = True)
        for room in warehouse:
            while i < len(boxes) and boxes[i] > room:
                i += 1
            if i == len(boxes):
                return count
            count += 1
            i += 1
        return count

File: src/leetcode_grind/day_578.rs
// https://leetcode.com/problems/put-boxes-into-the-warehouse-i/description/
pub fn max_boxes_in_warehouse(boxes: Vec<i32>, warehouse: Vec<i32>) -> i32 {
    let n = warehouse.len();
    let mut boxes = boxes;
    boxes.sort();
    let mut wh = vec![i32::MAX; warehouse.len()];
    wh[0] = warehouse[0];
    for i in 1..n {
        wh[i] = wh[i - 1].min(warehouse[i]);
    }
    let mut ans = 0;
    let mut j = 0;
    for i in (0..n).rev() {
        if j < boxes.len() && boxes[j] <= wh[i] {
            ans += 1;
            j += 1;
        }
    }

    ans
}
